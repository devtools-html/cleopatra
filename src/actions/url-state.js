/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type { Action } from '../types/store';

export function urlStateHasChanged(): Action {
  return {
    type: 'URL_STATE_HAS_CHANGED',
  };
}
