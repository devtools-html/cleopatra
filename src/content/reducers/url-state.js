// @flow
import { combineReducers } from 'redux';
import { defaultThreadOrder } from '../profile-data';
import { createSelector } from 'reselect';
import { urlFromState } from '../url-handling';
import * as RangeFilters from '../range-filters';

import type { ThreadIndex } from '../../common/types/profile';
import type { StartEndRange } from '../../common/types/units';
import type { Action, CallTreeFiltersPerThread, CallTreeFilter, DataSource } from '../actions/types';
import type { State, URLState, Reducer } from './types';

function dataSource(state: DataSource = 'none', action: Action) {
  switch (action.type) {
    case 'WAITING_FOR_PROFILE_FROM_FILE':
      return 'from-file';
    case 'PROFILE_PUBLISHED':
      return 'public';
    default:
      return state;
  }
}

function hash(state: string = '', action: Action) {
  switch (action.type) {
    case 'PROFILE_PUBLISHED':
      return action.hash;
    default:
      return state;
  }
}

function selectedTab(state: string = 'calltree', action: Action) {
  switch (action.type) {
    case 'CHANGE_SELECTED_TAB':
      return action.selectedTab;
    default:
      return state;
  }
}

function rangeFilters(state: StartEndRange[] = [], action: Action) {
  switch (action.type) {
    case 'ADD_RANGE_FILTER': {
      const { start, end } = action;
      return [...state, { start, end }];
    }
    case 'POP_RANGE_FILTERS':
      return state.slice(0, action.firstRemovedFilterIndex);
    default:
      return state;
  }
}

function selectedThread(state: ThreadIndex = 0, action: Action) {
  switch (action.type) {
    case 'CHANGE_SELECTED_THREAD':
      return action.selectedThread;
    case 'RECEIVE_PROFILE_FROM_ADDON':
    case 'RECEIVE_PROFILE_FROM_FILE': {
      // When loading in a brand new profile, select either the GeckoMain [tab] thread,
      // or the first thread in the thread order. For profiles from the Web, the
      // selectedThread has already been initialized from the URL and does not require
      // looking at the profile.
      const contentThreadId = action.profile.threads.findIndex(thread => {
        return thread.name === 'GeckoMain' && thread.processType === 'tab';
      });
      return contentThreadId !== -1 ? contentThreadId : defaultThreadOrder(action.profile.threads)[0];
    }
    default:
      return state;
  }
}

function callTreeSearchString(state: string = '', action: Action) {
  switch (action.type) {
    case 'CHANGE_CALL_TREE_SEARCH_STRING':
      return action.searchString;
    default:
      return state;
  }
}

function callTreeFilters(state: CallTreeFiltersPerThread = {}, action: Action) {
  switch (action.type) {
    case 'ADD_CALL_TREE_FILTER': {
      const { threadIndex, filter } = action;
      const oldFilters = state[threadIndex] || [];
      return Object.assign({}, state, {
        [threadIndex]: [...oldFilters, filter],
      });
    }
    case 'POP_CALL_TREE_FILTERS': {
      const { threadIndex, firstRemovedFilterIndex } = action;
      const oldFilters = state[threadIndex] || [];
      return Object.assign({}, state, {
        [threadIndex]: oldFilters.slice(0, firstRemovedFilterIndex),
      });
    }
    default:
      return state;
  }
}

function jsOnly(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'CHANGE_JS_ONLY':
      return action.jsOnly;
    default:
      return state;
  }
}

function invertCallstack(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'CHANGE_INVERT_CALLSTACK':
      return action.invertCallstack;
    default:
      return state;
  }
}

function hidePlatformDetails(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'CHANGE_HIDE_PLATFORM_DETAILS':
      return action.hidePlatformDetails;
    default:
      return state;
  }
}

const urlStateReducer: Reducer<URLState> = (regularUrlStateReducer => (state: URLState, action: Action): URLState => {
  switch (action.type) {
    case '@@urlenhancer/updateURLState':
      return action.urlState;
    default:
      return regularUrlStateReducer(state, action);
  }
})(combineReducers({
  dataSource, hash, selectedTab, rangeFilters, selectedThread,
  callTreeSearchString, callTreeFilters, jsOnly, invertCallstack,
  hidePlatformDetails,
}));
export default urlStateReducer;

const getURLState = (state: State): URLState => state.urlState;

export const getDataSource = (state: State) => getURLState(state).dataSource;
export const getHash = (state: State) => getURLState(state).hash;
export const getRangeFilters = (state: State) => getURLState(state).rangeFilters;
export const getJSOnly = (state: State) => getURLState(state).jsOnly;
export const getHidePlatformDetails = (state: State) => getURLState(state).hidePlatformDetails;
export const getInvertCallstack = (state: State) => getURLState(state).invertCallstack;
export const getSearchString = (state: State) => getURLState(state).callTreeSearchString;
export const getSelectedTab = (state: State) => getURLState(state).selectedTab;
export const getSelectedThreadIndex = (state: State) => getURLState(state).selectedThread;
export const getCallTreeFilters = (state: State, threadIndex: ThreadIndex): CallTreeFilter[] => {
  return getURLState(state).callTreeFilters[threadIndex] || [];
};

export const getURLPredictor = createSelector(
  getURLState,
  (oldURLState: URLState) => actionOrActionList => {
    const actionList = ('type' in actionOrActionList) ? [actionOrActionList] : actionOrActionList;
    const newURLState = actionList.reduce(urlStateReducer, oldURLState);
    return urlFromState(newURLState);
  }
);

export const getRangeFilterLabels = createSelector(
  getRangeFilters,
  RangeFilters.getRangeFilterLabels
);
