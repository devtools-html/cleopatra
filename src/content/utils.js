// @flow
type ValidWeakMapKey = Object | [];

/**
 * Memoize a function's results using a WeakMap and its first parameter.
 */
export function weakMemoize(fn: (ValidWeakMapKey, ...any) => any) {
  const memo = new WeakMap();
  return function (key: ValidWeakMapKey) {
    const cachedValue = memo.get(key);
    if (cachedValue === undefined) {
      const value = fn.apply(null, arguments);
      memo.set(key, value);
      return value;
    }
    return cachedValue;
  };
}
