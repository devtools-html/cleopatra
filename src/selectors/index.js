/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import { selectedThreadSelectors as selectedThread } from './per-thread';
import * as profile from './profile';
import * as urlState from './url-state';
import * as icons from './icons';
import * as publish from './publish';
import * as zippedProfiles from './zipped-profiles';

// Normally we don't need that type because flow can infer types, but on that
// object, it gives weird flow errors that say missing type annotation for X.
// Only adding a type like that seems to solve it.
type Selectors = {
  app: typeof app,
  profile: typeof profile,
  urlState: typeof urlState,
  icons: typeof icons,
  publish: typeof publish,
  zippedProfiles: typeof zippedProfiles,
  selectedThread: typeof selectedThread,
};

export default ({
  app,
  profile,
  urlState,
  icons,
  publish,
  zippedProfiles,
  selectedThread,
}: Selectors);
