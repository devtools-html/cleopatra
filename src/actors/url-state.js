/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import { getUrlState } from '../reducers/url-state';
import {
  urlStateHasChanged,
  updateUrlState,
  show404,
} from '../actions/url-state';
import { urlFromState, stateFromLocation } from '../url-handling';

import type { Dispatch } from '../types/store';
import type { State, UrlState } from '../types/reducers';

let previousUrlState = null;
export default function stateWatcher(state: State, dispatch: Dispatch) {
  function updateState() {
    if (window.history.state) {
      dispatch(updateUrlState(window.history.state));
    } else {
      try {
        const urlState = stateFromLocation(window.location);
        // replacing the existing location ensures we get the latest format for the URL
        updateLocation(urlState, true);
        dispatch(updateUrlState(urlState));
      } catch (e) {
        console.error(e);
        show404(window.location.pathname + window.location.search);
      }
    }
  }

  function initialSetup() {
    updateState();
    window.addEventListener('popstate', updateState);
  }

  function updateLocation(urlState: UrlState, replace: boolean = false) {
    const newUrl = urlFromState(urlState);
    const operation = replace
      ? window.history.replaceState.bind(window.history)
      : window.history.pushState.bind(window.history);
    if (newUrl !== window.location.pathname + window.location.search) {
      operation(urlState, document.title, newUrl);
    }
  }

  const newUrlState = getUrlState(state);
  if (!previousUrlState) {
    // first run
    previousUrlState = newUrlState;
    initialSetup();
  } else if (newUrlState !== previousUrlState) {
    previousUrlState = newUrlState;
    updateLocation(newUrlState);
    dispatch(urlStateHasChanged());
  }
}
