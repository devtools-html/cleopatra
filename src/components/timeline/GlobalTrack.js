/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import classNames from 'classnames';
import {
  changeRightClickedTrack,
  changeLocalTrackOrder,
  selectTrack,
} from '../../actions/profile-view';
import ContextMenuTrigger from '../shared/ContextMenuTrigger';
import {
  getSelectedThreadIndex,
  getHiddenGlobalTracks,
  getLocalTrackOrder,
  getSelectedTab,
} from '../../selectors/url-state';
import explicitConnect from '../../utils/connect';
import {
  getGlobalTracks,
  getLocalTracks,
  getGlobalTrackName,
  getProcessesWithMemoryTrack,
} from '../../selectors/profile';
import { getThreadSelectors } from '../../selectors/per-thread';
import './Track.css';
import TimelineTrackThread from './TrackThread';
import TimelineTrackScreenshots from './TrackScreenshots';
import TimelineLocalTrack from './LocalTrack';
import Reorderable from '../shared/Reorderable';
import type { TabSlug } from '../../app-logic/tabs-handling';
import type { GlobalTrackReference } from '../../types/actions';
import type { Pid } from '../../types/profile';
import type {
  TrackIndex,
  GlobalTrack,
  LocalTrack,
} from '../../types/profile-derived';
import type { ConnectedProps } from '../../utils/connect';

type OwnProps = {|
  +trackReference: GlobalTrackReference,
  +trackIndex: TrackIndex,
  +style?: Object /* This is used by Reorderable */,
|};

type StateProps = {|
  +trackName: string,
  +globalTrack: GlobalTrack,
  +isSelected: boolean,
  +isHidden: boolean,
  +titleText: string | null,
  +localTrackOrder: TrackIndex[],
  +localTracks: LocalTrack[],
  +pid: Pid | null,
  +selectedTab: TabSlug,
  +processesWithMemoryTrack: Set<Pid>,
|};

type DispatchProps = {|
  +changeRightClickedTrack: typeof changeRightClickedTrack,
  +changeLocalTrackOrder: typeof changeLocalTrackOrder,
  +selectTrack: typeof selectTrack,
|};

type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;

export const TRACK_PROCESS_BLANK_HEIGHT = 30;

class GlobalTrackComponent extends PureComponent<Props> {
  _onLabelMouseDown = (event: MouseEvent) => {
    const { changeRightClickedTrack, trackReference } = this.props;

    if (event.button === 0) {
      // Don't allow clicks on the threads list to steal focus from the tree view.
      event.preventDefault();
      this._selectCurrentTrack();
    } else if (event.button === 2) {
      // This is needed to allow the context menu to know what was right clicked without
      // actually changing the current selection.
      changeRightClickedTrack(trackReference);
    }
  };

  _selectCurrentTrack = () => {
    const { selectTrack, trackReference } = this.props;
    selectTrack(trackReference);
  };

  renderTrack() {
    const { globalTrack, processesWithMemoryTrack } = this.props;
    switch (globalTrack.type) {
      case 'process': {
        const { mainThreadIndex } = globalTrack;
        if (mainThreadIndex === null) {
          return (
            <div
              className="timelineTrackThreadBlank"
              style={{
                '--timeline-track-thread-blank-height': TRACK_PROCESS_BLANK_HEIGHT,
              }}
            />
          );
        }
        return (
          <TimelineTrackThread
            threadIndex={mainThreadIndex}
            showMemoryMarkers={!processesWithMemoryTrack.has(globalTrack.pid)}
          />
        );
      }
      case 'screenshots': {
        const { threadIndex, id } = globalTrack;
        return (
          <TimelineTrackScreenshots threadIndex={threadIndex} windowId={id} />
        );
      }
      default:
        console.error('Unhandled globalTrack type', (globalTrack: empty));
        return null;
    }
  }

  _changeLocalTrackOrder = (trackOrder: TrackIndex[]) => {
    const { globalTrack, changeLocalTrackOrder } = this.props;
    if (globalTrack.type === 'process') {
      // Only process tracks have local tracks.
      changeLocalTrackOrder(globalTrack.pid, trackOrder);
    }
  };

  renderLocalTracks(pid: Pid) {
    const { localTracks, localTrackOrder } = this.props;
    return (
      <Reorderable
        tagName="ol"
        className="timelineTrackLocalTracks"
        order={localTrackOrder}
        orient="vertical"
        grippyClassName="timelineTrackLocalGrippy"
        onChangeOrder={this._changeLocalTrackOrder}
      >
        {localTracks.map((localTrack, trackIndex) => (
          <TimelineLocalTrack
            key={trackIndex}
            pid={pid}
            localTrack={localTrack}
            trackIndex={trackIndex}
          />
        ))}
      </Reorderable>
    );
  }

  render() {
    const {
      isSelected,
      isHidden,
      titleText,
      trackName,
      style,
      localTracks,
      pid,
    } = this.props;

    if (isHidden) {
      // If this global track is hidden, render out a stub element so that the
      // Reorderable Component still works across all the tracks.
      return <li className="timelineTrackHidden" />;
    }

    return (
      <li className="timelineTrack" style={style}>
        <div
          className={classNames('timelineTrackRow timelineTrackGlobalRow', {
            selected: isSelected,
          })}
          onClick={this._selectCurrentTrack}
        >
          <ContextMenuTrigger
            id="TimelineTrackContextMenu"
            renderTag="div"
            attributes={{
              title: titleText,
              className: 'timelineTrackLabel timelineTrackGlobalGrippy',
              onMouseDown: this._onLabelMouseDown,
            }}
          >
            <button type="button" className="timelineTrackNameButton">
              {trackName}
              {
                // Only show the PID if it is a real number. A string PID is an
                // artificially generated value that is not useful, and a null
                // value does not exist. */
              }
              {typeof pid === 'number' ? (
                <div className="timelineTrackNameButtonAdditionalDetails">
                  PID: {pid}
                </div>
              ) : null}
            </button>
          </ContextMenuTrigger>
          <div className="timelineTrackTrack">{this.renderTrack()}</div>
        </div>
        {localTracks.length > 0 && pid !== null
          ? this.renderLocalTracks(pid)
          : null}
      </li>
    );
  }
}

// Provide some empty lists, so that strict equality checks work for component updates.
const EMPTY_TRACK_ORDER = [];
const EMPTY_LOCAL_TRACKS = [];

export default explicitConnect<OwnProps, StateProps, DispatchProps>({
  mapStateToProps: (state, { trackIndex }) => {
    const globalTracks = getGlobalTracks(state);
    const globalTrack = globalTracks[trackIndex];
    const selectedTab = getSelectedTab(state);

    // These get assigned based on the track type.
    let threadIndex = null;
    let isSelected = false;
    let titleText = null;

    let localTrackOrder = EMPTY_TRACK_ORDER;
    let localTracks = EMPTY_LOCAL_TRACKS;
    let pid = null;

    // Run different selectors based on the track type.
    switch (globalTrack.type) {
      case 'process': {
        // Look up the thread information for the process if it exists.
        if (globalTrack.mainThreadIndex !== null) {
          threadIndex = globalTrack.mainThreadIndex;
          const selectors = getThreadSelectors(threadIndex);
          isSelected =
            threadIndex === getSelectedThreadIndex(state) &&
            selectedTab !== 'network-chart';
          titleText = selectors.getThreadProcessDetails(state);
        }
        pid = globalTrack.pid;
        localTrackOrder = getLocalTrackOrder(state, pid);
        localTracks = getLocalTracks(state, pid);
        break;
      }
      case 'screenshots':
        break;
      default:
        throw new Error(`Unhandled GlobalTrack type ${(globalTrack: empty)}`);
    }

    return {
      trackName: getGlobalTrackName(state, trackIndex),
      titleText,
      globalTrack,
      isSelected,
      localTrackOrder,
      localTracks,
      pid,
      isHidden: getHiddenGlobalTracks(state).has(trackIndex),
      selectedTab,
      processesWithMemoryTrack: getProcessesWithMemoryTrack(state),
    };
  },
  mapDispatchToProps: {
    changeRightClickedTrack,
    changeLocalTrackOrder,
    selectTrack,
  },
  component: GlobalTrackComponent,
});
