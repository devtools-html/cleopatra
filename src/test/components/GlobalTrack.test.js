/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

import {
  changeSelectedThread,
  hideGlobalTrack,
} from '../../actions/profile-view';
import GlobalTrack from '../../components/timeline/GlobalTrack';
import { getGlobalTracks, getRightClickedTrack } from '../../selectors/profile';
import { getSelectedThreadIndex } from '../../selectors/url-state';
import mockCanvasContext from '../fixtures/mocks/canvas-context';
import { getProfileWithNiceTracks } from '../fixtures/profiles/tracks';
import { storeWithProfile } from '../fixtures/stores';
import { getMouseEvent } from '../fixtures/utils';

const LEFT_CLICK = 0;
const RIGHT_CLICK = 2;

describe('timeline/GlobalTrack', function() {
  /**
   *  getProfileWithNiceTracks() looks like: [
   *    'show [thread GeckoMain process]',
   *    'show [thread GeckoMain tab]',       <- use this global track.
   *    '  - show [thread DOM Worker]',
   *    '  - show [thread Style]',
   *  ]
   */
  function setup() {
    const trackIndex = 1;
    const profile = getProfileWithNiceTracks();
    const store = storeWithProfile(profile);
    const { getState, dispatch } = store;
    const trackReference = { type: 'global', trackIndex };
    const tracks = getGlobalTracks(getState());
    const track = tracks[trackIndex];
    if (track.type !== 'process') {
      throw new Error('Expected a process track.');
    }
    const threadIndex = track.mainThreadIndex;
    if (threadIndex === null) {
      throw new Error('Expected the track to have a thread index.');
    }

    // Some child components render to canvas.
    jest
      .spyOn(HTMLCanvasElement.prototype, 'getContext')
      .mockImplementation(() => mockCanvasContext());

    // The assertions are simpler if this thread is not already selected.
    dispatch(changeSelectedThread(threadIndex + 1));

    const view = mount(
      <Provider store={store}>
        <GlobalTrack trackIndex={trackIndex} trackReference={trackReference} />
      </Provider>
    );

    const getGlobalTrackLabel = () => view.find('.timelineTrackLabel').first();
    const getGlobalTrackRow = () =>
      view.find('.timelineTrackGlobalRow').first();

    return {
      dispatch,
      getState,
      profile,
      store,
      view,
      trackReference,
      trackIndex,
      threadIndex,
      getGlobalTrackLabel,
      getGlobalTrackRow,
    };
  }

  it('matches the snapshot of a global track', () => {
    const { view } = setup();
    expect(view).toMatchSnapshot();
  });

  it('has the correct selectors into useful parts of the component', function() {
    const { getGlobalTrackLabel, getGlobalTrackRow } = setup();
    expect(getGlobalTrackLabel().text()).toBe('Content Process');
    expect(getGlobalTrackRow().exists()).toBe(true);
  });

  it('starts out not being selected', function() {
    const {
      getState,
      getGlobalTrackRow,
      threadIndex,
      trackReference,
    } = setup();
    expect(getRightClickedTrack(getState())).not.toEqual(trackReference);
    expect(getSelectedThreadIndex(getState())).not.toBe(threadIndex);
    expect(getGlobalTrackRow().hasClass('selected')).toBe(false);
  });

  it('can select a thread by clicking the label', () => {
    const {
      getState,
      getGlobalTrackLabel,
      getGlobalTrackRow,
      threadIndex,
    } = setup();
    expect(getSelectedThreadIndex(getState())).not.toBe(threadIndex);
    getGlobalTrackLabel().simulate(
      'mousedown',
      getMouseEvent({ button: LEFT_CLICK })
    );
    expect(getSelectedThreadIndex(getState())).toBe(threadIndex);
    expect(getGlobalTrackRow().hasClass('selected')).toBe(true);
  });

  it('can right click a thread', () => {
    const {
      getState,
      getGlobalTrackLabel,
      threadIndex,
      trackReference,
    } = setup();

    getGlobalTrackLabel().simulate(
      'mousedown',
      getMouseEvent({ button: RIGHT_CLICK })
    );
    expect(getRightClickedTrack(getState())).toEqual(trackReference);
    expect(getSelectedThreadIndex(getState())).not.toBe(threadIndex);
  });

  it('can select a thread by clicking the row', () => {
    const { getState, getGlobalTrackRow, threadIndex } = setup();
    expect(getSelectedThreadIndex(getState())).not.toBe(threadIndex);
    getGlobalTrackRow().simulate('click');
    expect(getSelectedThreadIndex(getState())).toBe(threadIndex);
  });

  it('will render a stub div if the track is hidden', () => {
    const { view, trackIndex, dispatch } = setup();
    dispatch(hideGlobalTrack(trackIndex));
    view.update();
    expect(view.find('.timelineTrackHidden').exists()).toBe(true);
    expect(view.find('.timelineTrack').exists()).toBe(false);
  });
});
