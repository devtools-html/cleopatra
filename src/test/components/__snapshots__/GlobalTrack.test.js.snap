// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeline/GlobalTrack matches the snapshot of a global track 1`] = `
<li
  className="timelineTrack"
>
  <div
    className="timelineTrackRow timelineTrackGlobalRow"
    onClick={[Function]}
  >
    <div
      className="react-contextmenu-wrapper timelineTrackLabel timelineTrackGlobalGrippy"
      onContextMenu={[Function]}
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      title="thread: \\"GeckoMain\\" (0)
process: \\"tab\\" (222)"
    >
      <button
        className="timelineTrackNameButton"
        type="button"
      >
        Content Process
      </button>
    </div>
    <div
      className="timelineTrackTrack"
    >
      <div
        className="timelineTrackThread"
      >
        <div
          className="timelineTrackThreadIntervalMarkerOverview"
        >
          <canvas
            className="timelineTracingMarkersCanvas"
            onMouseDown={[Function]}
            onMouseMove={[Function]}
            onMouseOut={[Function]}
            onMouseUp={[Function]}
          />
        </div>
        <div
          className="
              timelineTrackThreadIntervalMarkerOverview
              timelineTrackThreadIntervalMarkerOverviewThreadGeckoMain
            "
        >
          <canvas
            className="timelineTracingMarkersCanvas"
            onMouseDown={[Function]}
            onMouseMove={[Function]}
            onMouseOut={[Function]}
            onMouseUp={[Function]}
          />
        </div>
        <div
          className="threadActivityGraph"
        >
          <canvas
            className="threadActivityGraphCanvas threadActivityGraphCanvas"
            onMouseUp={[Function]}
          />
        </div>
        <div
          className="timelineEmptyThreadIndicator"
        />
      </div>
    </div>
  </div>
  <ol
    className="timelineTrackLocalTracks"
    onMouseDown={[Function]}
  >
    <li
      className="timelineTrack timelineTrackLocal"
    >
      <div
        className="timelineTrackRow timelineTrackLocalRow selected"
        onClick={[Function]}
      >
        <div
          className="react-contextmenu-wrapper timelineTrackLabel timelineTrackLocalLabel timelineTrackLocalGrippy"
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          onMouseOut={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          title="thread: \\"DOM Worker\\" (0)
process: \\"tab\\" (222)"
        >
          <button
            className="timelineTrackNameButton"
            type="button"
          >
            DOM Worker
          </button>
        </div>
        <div
          className="timelineTrackTrack"
        >
          <div
            className="timelineTrackThread"
          >
            <div
              className="timelineTrackThreadIntervalMarkerOverview selected"
            >
              <canvas
                className="timelineTracingMarkersCanvas"
                onMouseDown={[Function]}
                onMouseMove={[Function]}
                onMouseOut={[Function]}
                onMouseUp={[Function]}
              />
            </div>
            <div
              className="threadActivityGraph"
            >
              <canvas
                className="threadActivityGraphCanvas threadActivityGraphCanvas"
                onMouseUp={[Function]}
              />
            </div>
            <div
              className="timelineEmptyThreadIndicator"
            />
          </div>
        </div>
      </div>
    </li>
    <li
      className="timelineTrack timelineTrackLocal"
    >
      <div
        className="timelineTrackRow timelineTrackLocalRow"
        onClick={[Function]}
      >
        <div
          className="react-contextmenu-wrapper timelineTrackLabel timelineTrackLocalLabel timelineTrackLocalGrippy"
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          onMouseOut={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          title="thread: \\"Style\\" (0)
process: \\"tab\\" (222)"
        >
          <button
            className="timelineTrackNameButton"
            type="button"
          >
            Style
          </button>
        </div>
        <div
          className="timelineTrackTrack"
        >
          <div
            className="timelineTrackThread"
          >
            <div
              className="timelineTrackThreadIntervalMarkerOverview"
            >
              <canvas
                className="timelineTracingMarkersCanvas"
                onMouseDown={[Function]}
                onMouseMove={[Function]}
                onMouseOut={[Function]}
                onMouseUp={[Function]}
              />
            </div>
            <div
              className="threadActivityGraph"
            >
              <canvas
                className="threadActivityGraphCanvas threadActivityGraphCanvas"
                onMouseUp={[Function]}
              />
            </div>
            <div
              className="timelineEmptyThreadIndicator"
            />
          </div>
        </div>
      </div>
    </li>
  </ol>
</li>
`;
